plugins/ports-plugin/src/endpoints-tree-data-provider.ts

```diff
@@ -0,0 +1,245 @@
+/*********************************************************************
+ * Copyright (c) 2020 Red Hat, Inc.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ **********************************************************************/
+
+import * as theia from '@theia/plugin';
+import { Endpoint } from './endpoint';
+import { EndpointCategory } from './endpoint-category';
+import { EndpointExposure } from './endpoint-exposure';
+import { ListeningPort } from './listening-port';
+
+// available context values used in package.json to assign context menus
+export type EndpointTreeNodeItemContext = 'publicHttpsEndpointOnline' | 'publicHttpPortOnline' | 'publicPortOnline' | 'publicDevfilePortOffline'
+    | 'privateUserPortOnline' | 'privateDevfilePortOnline' | 'privateDevfilePortOffline';
+
+// defines a custom item by adding the endpoint and parent id.
+export interface EndpointTreeNodeItem extends theia.TreeItem {
+    endpoint?: Endpoint;
```

Comment:
Can we override the `contextValue` property to be of type `EndpointTreeNodeItemContext` in order to have to have type safety?